src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java

```diff
@@ -2849,26 +2842,55 @@ public void watchFiles(File outputDirectory, File testOutputDirectory, final Thr
                             testArtifactPaths, null, false);
                 }
 
-                // when change in class files are detected scan for Liberty features.
-                if (generateFeatures && !javaSourceClasses.isEmpty()) {
-                    debug("Detected a change in the following classes: " + javaSourceClasses);
-                    if (!gradle && !foundInitialClasses) {
-                        // For Maven, skip the first call of generate features from the initial Java compilation that happens after dev mode startup
-                        // because features were already generated during the actual dev mode startup steps.
-                        debug("Skipping generate features from first call after dev mode startup");
-                        foundInitialClasses = true;
-                        javaSourceClasses.clear();
-                    } else if (!classesFailingToCompile()){ // do not run generate features if there are classes failing to compile
-                        incrementGenerateFeatures();
-                        if (!generatedFeaturesFile.exists()) {
-                            // run tests if generated-features.xml does not exist as there are no new features to install
-                            if (isMultiModuleProject()) {
-                                runTestThread(false, executor, -1, false, getAllBuildFiles());
-                            } else {
-                                runTestThread(false, executor, -1, false, false, buildFile);
+                // Assumption is that initialCompile will only be set to false after ALL modules are complete
+                if (!recompileDependencies && !registerClassDirAttempted && !initialCompile) {
+                    debug("Attempt to register class file directories for feature generation");
+                    // set to true regardless of success
+                    // this may need to be updated, but currently do not want to come into this section again
+                    registerClassDirAttempted = true;
```

Comment:
Why do we only want to enter this block once? Will we catch the case where java classes are added after dev mode runs for a while, and then we register the output directory?

Rather than the `registerClassDirAttempted` boolean, instead can we use the `sourceOutputDirRegistered` existing booleans similar to how we do for `sourceDirRegistered` once we are in the watch loop?
ie. 
```java
if (!recompileDependencies && !initialCompile){
    if (isMultiModuleProject()) {
        for (ProjectModule p : upstreamProjects) {
              if (!p.sourceOutputDirRegistered && p.getOutputDirectory.exists() && shouldIncludeSources(p.getPackagingType())) {
                  // register  
               }
        }
    }
    if (shouldIncludeSources(packagingType) && !sourceOutputDirRegistered && this.outputDirectory.exists()) {
        // register 
    }
}
```

Reply:
Updated to only have one block where we attempt to register the output directories.